pkg/util/watch/relist.go

```diff
@@ -58,6 +81,51 @@ func (r *relistInfo) Relist() <-chan struct{} {
 	return r.relisted
 }
 
+var closedChan = makeClosedChan()
+
+func makeClosedChan() <-chan struct{} {
+	c := make(chan struct{})
+	close(c)
+	return c
+}
+
+func (r *relistInfo) RelistIfHaventSince(c RelistCount) <-chan struct{} {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if c < r.relistCount {
+		return closedChan
+	}
+
+	upperBound := r.lockedRelistCountUpperBound()
+	// we know c >= r.relistCount, so if c < upperBound we know that the higher upper bound comes
+	// from the executor.
+	if c < upperBound {
+		// doesn't matter which channel we return (they'll all get closed at once)
+		// So it's easiest just to return the first.
+		return r.executor.signalComplete[0]
+	}
+
+	// Otherwise, the relist hasn't already happened, and there isn't one in progress that will
+	// satisfy what we're looking for, so we need to start a fresh one.
+	return r.lockedRelist()
+}
+
+func (r *relistInfo) RelistCountUpperBound() RelistCount {
```

Comment:
Upper bound exists because it includes relists that may not have completed yet. The tl;dr on why is that having a "single" RelistCount value (instead of explicitly only giving an upper bound) is not possible to make sound. I'm struggling to phrase the reasoning well - I'll write a follow-up explaining it in more detail when I get the chance.

Reply:
Ok, best attempt. LMK if this makes sense:

So if we returned a "single" relist count (i.e. at each moment in time, there is a defined value equal to the number of relists that have occurred, and we provide that value), there's two ways we could do it.

1. Return the number of *successful* relists (i.e. only increment once `List()` is successful & fully processed); or
2. Return the number of *attempted* relists (i.e. increment each time relisting *starts*)

In both cases, it's possible to get unsound behavior through reasonable usage of the API.

Let's pretend we have two threads, 'A' (will receive events & fetch relist count) and 'B' (will perform relisting).

For the first case (increment only once relist is complete), we can get unsound behavior with the following sequence of events:

```mermaid
sequenceDiagram

    participant k8s as K8s API server
    participant a as Thread A
    participant relistCount
    participant b as Thread B

    note over relistCount: initial value = 0

    note over b: Relisting starts<br>(requested externally)
    b->>k8s: List()
    k8s->>b: List() returns
    note over b: List events processed

    note over k8s: New event generated for A
    k8s->>a: Event received via Watch() stream
    a->>relistCount: Fetch relist count
    relistCount->>a: relistCount = 0

    b->>relistCount: relistCount += 1

    a->>relistCount: RelistIfHaventSince(0)
    relistCount->>a: immediately returns

    note over a: error: RelistIfHaventSince(..) should have relisted<br>because event is newer than List() contents
```

For the second case (increment when relisting starts) - this is close to what we have. The only difference is that for the PR in its current state, we call this the "upper bound". If we *didn't*, then someone might assume comparison between `RelistCount` is valid, which it isn't. For example:

```mermaid
sequenceDiagram

    participant k8s as K8s API server
    participant a as Thread A
    participant relistCount
    participant b as Thread B

    note over k8s: New event generated for A
    k8s->>a: Event received via Watch() stream

    a->>relistCount: Fetch relist count
    relistCount->>a: relistCount = 0
    note over a: 'A' now knows that the event<br>happens-before relistCount > 0<br>... right?

    note over b: Relisting starts<br>(requested externally)
    b->>relistCount: relistCount += 1

    a->>relistCount: Fetch relist count
    relistCount->>a: relistCount = 1
    note over a: error: 'A' assumes contents of the store<br>must be newer than the event

    b->>k8s: List()
    k8s->>b: List() returns
    note over b: List events processed
```

Maybe this is ok! I think it is, as long as we make it explicit that the `RelistCount` you get back is just an upper bound (and so, can't really be compared).

Reply:
Thanks for the diagrams! Super helpful!

Definitely we should go with the second definition as you said, since the first one is not sound.

I understood the error you pointed out, and it took me a little bit to get it. I think there is a component missing in this discussion, which is the store. The store also has a `relistCount`, which is indicates the `relistCount` of the last relist operation that finished and updated the store's content. So in the case above, Thread A should actually compare its `relistCount` to the store's `relistCount`, which in this case there wouldn't be an error since both are equal to zero.

Considering store's `relistCount` explicitly, I understood why you used the term upper-bound since store's `relistCount` lags behind because we might have operations that aren't finished yet (or even failed).

I think there are 3 main entities here that each have their own `relistCount`:
- Relister: This is the component who is responsible for doing the relist operations. I think this is equivalent to thread B in your description. Relister has a `relistCount`, which indicated the number of attempted relists (we need to update this just before doing each relist op).
- Store: the VM store. It has a `relistCount` that corresponds to the last successful relist operation.
- Pod Event: It has a `relistCount` that shows the number of relist operations that have attempted before receiving this pod event.

What do you think of this breakdown here? Do you think it makes sense to add `relistCount` to Store? I think this way we could be more precise about the semantics.

Also thinking of each relist operation separately and its associated relist count is very helpful to reason the entire design.

Reply:
Also, I made a diagram while thinking about this and here I'm sharing it with you because it looks nice lol.
<img width="984" alt="Screenshot 2024-03-12 at 2 36 39 PM" src="https://github.com/neondatabase/autoscaling/assets/1623278/c40e10b5-9972-4963-98cf-b075e3124f41">

This figure summarizes what we want to do with this PR. We want to avoid performing relist for the blue events, since they could use the result from the relist that we do for the red event.